# 3.2 Bucles
***

Hemos hablado anteriormente acerca de los condicionales, que nos permiten ejecutar diferentes fragmentos de código en función de ciertas condiciones. Ahora, nos centraremos en otra estructura de control, los **bucles**. Los bucles nos ofrecen la posibilidad de repetir un mismo fragmento de código un número específico de veces o mientras se cumpla una determinada condición. Los bucles son una herramienta poderosa y versátil en la programación, ya que nos permiten automatizar tareas y evitar la repetición manual del mismo código una y otra vez.

Existen dos tipos principales de bucles en Python: el bucle `for` y el bucle `while`. El bucle `for` nos permite iterar sobre una secuencia de elementos, como una lista, una cadena de texto o un rango de números, y ejecutar el bloque de código correspondiente para cada elemento de la secuencia. Por otro lado, el bucle `while` se ejecuta mientras se cumpla una determinada condición, y nos permite repetir un bloque de código hasta que dicha condición deje de ser verdadera.

```{note} 
Podemos clasificar los bucles en dos categorías principales según su forma de ejecución: bucles determinados (`for`) y bucles indeterminados (`while`). 
```

Una de las principales ventajas de los bucles es que nos permiten ahorrar código y simplificar tareas repetitivas. En lugar de escribir instrucciones manualmente para cada repetición, podemos utilizar un bucle para automatizar la ejecución del código. Esto no solo ahorra tiempo y esfuerzo, sino que también facilita el mantenimiento y la modificación del código en el futuro, ya que solo tenemos que realizar cambios en un único bloque de código en lugar de múltiples repeticiones dispersas.

Los bucles son especialmente útiles cuando necesitamos procesar grandes volúmenes de datos, realizar cálculos repetitivos, interactuar con elementos de una lista o realizar cualquier tarea que requiera repeticiones controladas. Su flexibilidad y capacidad para adaptarse a diferentes situaciones los convierten en una herramienta esencial en la programación, permitiéndonos escribir código más eficiente y conciso.


## Ciclo `for`
***

En Python, el bucle `for` se utiliza como una forma general de iterar sobre una secuencia, es decir, recorre los elementos de un objeto iterable (listas, tuplas, diccionarios, entre otros) y ejecuta un bloque de código. En cada paso de la iteración considera a un único elemento del objeto iterable, sobre el cuál se pueden aplicar una serie de operaciones.

La estructura del ciclo for, es la siguiente:

```python
for var in objeto:
    
    Instrucción
```

donde:
    
- `var` → Representa la variable que se va iterando .
- `in` → Representa el operador que permite evaluar si el valor especificado se enceuntra dentro de la secuencia.
- `objeto` → Elemento implementado dentro de la estructura para realizar la indentación.
- `Instrucción` → Representa las instrucciones que se van a repetir cierto número de veces. 

### Ejemplo 1: Listas
 
 Para el siguiente ejemplo se hace uso de una lista

In [ ]:
Numeros = [1,2,3]

for i in Numeros: 
    
    print(i)
    
print("\n") 

De acuerdo a la lista definida como `Numeros`, el cual contiene en ella números que esta determinados, por lo que es posible recorrer mediante el ciclo **for**. La variable que se usa en este ejemplo es: `i`, el cual iterarará en la lista `Numeros`. 

El `print(i)` la función que cumple, es que cada vez que entre el **bucle for**, será un elemento de la lista de `Numeros`. Por esta razón en la primera iteración del bucle elemento valdrá `1`, en la segunda `2`, y en la tercera `3`.

|Estructura| Estructura|
| ---|---|
| "Variable" | `i` |
| "Operador" | `in`|
| "Objeto" | `Numeros`|
| "Instrucción" | `print("i")`|



En el siguiente ejemplo, se hace uso de una lista que contiene palabras

In [ ]:
for palabra in ["Hola", "Python", "FCFM", "Luis"]:  
    
    print(palabra, end=" ")

|Estructura| Estructura|
| ---|---|
| "Variable" | `palabra` |
| "Operador" | `in`|
| "Objeto" | `["Hola", "Python", "FCFM", "Luis"]`|
| "Instrucción" | `print(palabra, end=" ")`|

En este ejemplo, ocurre lo mismo que en el anterior, con la diferencia que lo que se itera son palabras. 

**Nota**
- *`end=" "`* → Representa un espacio en cada palabra cada vez que el ciclo for este iterando, es decir las palabras ya no estarán separadas con un salto de renglón.

## Range 

Anteriormente se trabajó con una lista que iteraba sobre sus elementos, sin embargo, esto representa una acción ineficiente ya que pueden existir listas que contengan un mayor número de elementos. Por ejemplo, una lista que requiera añadir números del `0` al `1000`, esta acción se volvería tediosa. 

Para ello, el lenguaje `Python`, aporta una función llamada **range** que permite generar una lista que vaya desde el primer número que le indiquemos al segundo número que se quiera. 


Previamente se uso este ejemplo, con el bucle **for** para imprimir los elementos de la lista `Numeros`, ahora si se requiere iterar hasta un delimitado número de elementos, se hace uso del "range". 

```{warning}
El range comienza desde una poscición ya establecida en la posición 0, por lo tanto el bucle itera hasta el número que se asigna, menos uno. 
```

La función range abarca mas beneficios para iterar sobre un bucle for, ya que si se desea eleguir el elemento en el que se quiere comenzar hasta que elemento se quiere terminar y establecer una secuencia entre el elemento inicial hasta el elemento de llegada, se agrega una comas más, tal como muestra el siguiente formato: `range(Inicio, Final, Salto)`

Donde:
    
- *`Inicio`* → Representa el elemento donde comienza el bucle .
- *`Final`* → Representa el elemento donde termina el bucle .
- *`salto`* → Representa la secuencia entre elemento inicial y final
`

Para que quede más claro, se toma el siguiente ejemplo: 

In [8]:
for i in range(7, 12, 2):
    
    print(i)
    

7
9
11


En este caso, el ciclo for recorre hasta el elemento `12` en un salto de dos en dos. 

Otra implementación que agrega el "range" es la transformación de listas a tuplas, es decir transformar el rango en lista. 
Para ello se toma el siguiente ejemplo:

In [ ]:
x = list(range(1, 101))

print(x)

De acuerdo a esta ejemplo, se define una variable `x` que imprime una lista que va desde el `1` al `100`, pero para realizar la transformación a lista se antepone la palabra `list` seguido de la función `range`, de esta forma no es necesario definir una lista previamente que contenga una cantidad elementos tan grande ya que se puede crear gracias a la función "range". 



## Iteración sobre cadenas de texto

`Python`, también nos brinda la iteración sobre cadenas de cáracteres, esto se debe a que `Python` interpreta las cadenas de texto como una `lista de letras`. 

Para introducir lo anterior, se toma el siguiente ejemplo: 

In [ ]:
H = "Hola mundo, estamos aprendiendo bucles"

Se define una variable `H`, en donde de acuerdo a lo que contiene la cadena de texo, se puede decir que el primer elemento de esta lista es `H`, el segundo elemento es `o`, el tercer elemento es `l`, el cuarto es `a`, el quinto es `un espacio`, etc. 

Ahora, se hace uso de un ciclo for, que itera sobre la variable `H`.

In [ ]:
H = "Hola mundo, estamos aprendiendo bucles"
for letra in H:
    print(letra, end="")

De acuerdo al ciclo for, donde: 

|Estructura| Estructura|
| ---|---|
| "Variable" | `letra` |
| "Operador" | `in`|
| "Objeto" | `H`|
| "Instrucción" | `print(letra, end="")`|

Lo que hace es iterar sobre cada elemento de la lista, por lo que al imprimir el resultado arrojado será: `Hola mundo, estamos aprendiendo bucles`.

Sin embargo, hay distintas maneras de poder imprimir esta variable `H`. El sigueinte ejemplo muestra un formato que empezó a estar en `python 3.6`, que es utilizar una letra `f` al principio de la cadena que se quiera imprimir, luego se abren llaves para imprimir la variable, es decir: 
`print(f"Cadena de texto: {variable}")`. 

Utilizando el formato `f`, se utiliza el siguiente ejemplo. 



In [11]:
for k in range(11):
    print(f"El número es {k}")    
    

    

El número es 0
El número es 1
El número es 2
El número es 3
El número es 4
El número es 5
El número es 6
El número es 7
El número es 8
El número es 9
El número es 10


NameError: name 'Numeros' is not defined

|Estructura| Estructura|
| ---|---|
| "Variable" | `K` |
| "Operador" | `in`|
| "Objeto" | `[0,1,2,3,4,5,6,7,8,9,10,11]`|
| "Instrucción" | `print(f"El número es {k}")`|

El ciclo for recorre un rango que abarca hasta el elemento `10`, lo que ocurre es que a cada elemento de ese rango al imprimir realiza un salto de linea añadiendo `El número es`, es decir: 
Para el el elemento `1`, imprime `El número es 0`, para el elemento `2`, `El número es 1`, etc.

Para terminar con el bucle **for**, para entender la utilidad. En ocasiones se quiere evaluar, sumas o mejor conocidas como sumatorias de un conjunto de numeros. 

Para ello, imagine que desea sumar un conunto de números enteros que abarca hasta el `100`. Con el bucle **for** esta acción se vulve fácil ya que 

In [15]:
suma = 0 
for i in range(101): 
    suma += i 
print("La suma de los primeros 100 enteros es: ", suma)

La suma de los primeros 100 enteros es:  5050


Donde: 
- *`+=`* → suma = suma + i .

Para ello se requiere definir la variable `suma`, ya que la máquina necesita intepretar que es a lo que se refiere con suma. 
Para ello se define `suma` fuera del ciclo for, con valor `0`. Esto significa que dado que `suma` vale `0` no alterara el resultado ya que al entrar al bucle se sumará con el primer elemento que recorre la variable `i` que es `0`, pero `i` aumentará en cada iteración. 

Dando como resultado: `5050`

Esta sumatoria puede resultar un tanto trivial, dado que ya existe una fomrula la cual nos dice que si se quiere sumar hasta `n` elementos, entonces la suma es igual a: $$suma = \frac{ n(n+1)}{2}$$

## Bucle While 

Acontinuación se hablará de lo que es un ciclo **while**, anteriormente se explicó lo que era el ciclo **for** el cual se utilizaba para iterar sobre un cierto número de veces `determinado` , en cambio el bucle **while** es un ciclo `indeterminado`, es decir, que no se conoce el número de veces que se va ejecutar la instrucción que esta contenida dentro del ciclo while. 

La estructura del ciclo while, es la siguiente: 


In [17]:
while conidición: 
    Instrucción 

NameError: name 'conidición' is not defined

El ciclo while significa que mientras se cumpla dicha condición se ejecutará la instrucción. 

In [ ]:
Para ello se pasará al siguiente ejemplo: 

In [23]:
i = 0
while i <= 10: 
    print(i)
    i = i + 1 
    

0
1
2
3
4
5
6
7
8
9
10


La manera de iterar en un ciclo while es que se debe definir una variable de iteración, en este caso `i = 0` en donde la condición que `i <= 10`, imprime la variable `i`, sin embargo esta variable `i`, es una variable que se mantiene estática a diferencia del ciclo que **for** que esta en constante aumento. Si esta variable `i` no se aumenta se volverá un ciclo infinito porque todo el tiempo `i` es igual a `0`.  Para ello se agrega la instrucción que `i = i + 1`, en este caso la variable `i` entra al ciclo **while** como `0`, luego imprime la variable `i` para después aumenatrse y de esa forma se evita que la variable sea estática para que esta comience a reasignarse hasta que cumpla la condición `i <= 10`.  

Para el siguiente ejemplo, se utiliza el anterior añadiendo en este caso la orden **break** 

In [ ]:
i = 0
while i <= 10: 
    if i == 5:
        print("ciclo terminado")
        break
    print(i)
    i = i + 1 

El **break**, se usa para detener el ciclo cuando la variable `i` llegue a un cierto número que este propuesto por un condicional, tal fue el caso del `if i == 5:`,es decir, que sale del bucle para ejecutar las demas instrucciones si es que hay alguna otra, tal como se muestra en el ejemplo, en donde la primera iteración es `0`, la segunda itearción `1`, tercera iteración `2`, cuarta iteración `3` y quinta iteración `4`; dado que se cumple la condición que `i == 5` ejecuta la condición imprimiendo `ciclo terminado` y corta el ciclo por lo que ya no realiza las demas iteraciones.

También se encuentra la declaración `continue`, su función es saltarse la iteración, para regresar al bucle **while** y continuar ejecutandolo. Como se muestra en el ejemplo.

In [38]:
i = 0

while i <= 10: 
    
    i = i + 1 
    
    if i == 5:
        
        print("Nos saltamos el 5")
        
        continue
    print(i)

1
2
3
4
Nos saltamos el 5
6
7
8
9
10
11


## Ejemplo con ciclo while, "Determinar números pares o impares"
Para el siguiente ejemplo se basa en determinar cuales números son pares o impares. 




In [ ]:
N = int(input("Dame el número que deseas: "))

i = 1 

#Operación módulo %   a%b

while i <= N:
    
    if i%2 == 0:
        
        print(f"{i} es un número par")
        
    else:
        
        print(f"{i} es un número impar")
        
    i += 1

Para ello, se pide al usuario hasta que número se quiere obtener los resultados. 

Par ello se usa un cicilo **while**, en el cual previamente se define una variable de iteración como `i = 0`. La condición del ciclo **while** esta dada por: `i <= N`, iterará sobre todos esos números. Luego se añade una condicional para determinar si la variable de iteración es par o impar, para ello se usa la operación módulo que esencia es hacer la división de esos dos números y obtener el resto. 

- `si el módulo es 0, entonces es par`
- `si el módulo es distinto de 0, entonces es impar`

Siguiendo estos punto se relizan las condiciones, en donde si `if i%2 == 0: ` entonces es par, y la opción por default que viene dada por `else` que significa que el número es impar. 

Recordando que la variable `i` debe estar aumentado y asi evitar un ciclo infinto 

## While True
Otra forma de realizar un ciclo **while**, es utilizando la variable `True`, dado que `True` es verdadero, es una condición que siempre se cumple por lo que el ciclo comenzará a iterarse de manera indefinida, donde la única manera de cortarse es por medio de un **break**. 

Para ello se propone el siguiente ejemplo: 

In [ ]:
while True: 
    
    edad = int(input("Digita tu edad: "))
    
    if edad == 18 and edad >= 18:
        
        print("Acceso a la página")
        
        break 
    
    else:
    
        print("No tienes acceso")
        print("Eres menor de edad")

En este programa, se pide al usuario ingresra su edad. Luego se hace uso de un condicional `if edad >= 0 and edad >= 18:`

Notese que el usario puede digitar las veces que desee la edad, aun si la condición no es valida, ya que el bucle seguirá repitiendose hasta que la condición se cumpla y logre acceder a la página